source/adapters/level_zero/enqueue_native.cpp

```diff
@@ -10,11 +10,64 @@
 
 #include <ur_api.h>
 
+#include "logger/ur_logger.hpp"
 #include "queue.hpp"
+#include "ur_level_zero.hpp"
 
 ur_result_t ur_queue_handle_legacy_t_::enqueueNativeCommandExp(
-    ur_exp_enqueue_native_command_function_t, void *, uint32_t,
-    const ur_mem_handle_t *, const ur_exp_enqueue_native_command_properties_t *,
-    uint32_t, const ur_event_handle_t *, ur_event_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_exp_enqueue_native_command_function_t pfnNativeEnqueue, void *data,
+    uint32_t, const ur_mem_handle_t *,
+    const ur_exp_enqueue_native_command_properties_t *,
+    uint32_t NumEventsInWaitList, const ur_event_handle_t *phEventList,
+    ur_event_handle_t *phEvent) {
+  auto Queue = this;
+
+  // Lock automatically releases when this goes out of scope.
+  std::scoped_lock<ur_shared_mutex> lock(Queue->Mutex);
```

Comment:
This will deadlock. Similarly to how you did in in CUDA, we need a thread-local variable that will store `ZeCommandList` and return it in https://github.com/oneapi-src/unified-runtime/blob/main/source/adapters/level_zero/queue.cpp#L701 if its not null.

Reply:
Thanks @pbalcer have updated. Let me know if you think this will work.